File: demo/template.py
"""Mirrors CVE-2019-10906: Jinja2 sandbox escape + classic SSTI / XSS.

Weakness description: "Server-side template injection via user-controlled
template string" and "Cross-site scripting via unescaped output".
"""
from jinja2 import Environment, Template


# VULNERABLE: user `bio` is concatenated into the template SOURCE, not data —
# so {{ 7*7 }} or {{ config.items() }} executes server-side (SSTI).
def render_profile(username, bio):
    source = f"""
    <html>
      <body>
        <h1>Profile: {username}</h1>
        <div class="bio">{bio}</div>
      </body>
    </html>
    """
    return Template(source).render()


# VULNERABLE: autoescape disabled, raw user content rendered as HTML (XSS)
_env = Environment(autoescape=False)


def render_comment(author, comment):
    tmpl = _env.from_string("<div><strong>{{ author }}</strong>: {{ comment | safe }}</div>")
    return tmpl.render(author=author, comment=comment)


# SAFE reference (after fix):
# _env_safe = Environment(autoescape=True)
# def render_profile_safe(username, bio):
#     tmpl = _env_safe.from_string("<h1>Profile: {{ username }}</h1><div>{{ bio }}</div>")
#     return tmpl.render(username=username, bio=bio)

File: demo/src/template.js
// Mirrors CVE-2017-0931: sanitize-html — XSS via insufficient sanitisation
// Weakness description: "Cross-site scripting via unsanitized HTML output"
// Vulnerable: user-supplied bio is passed through an old sanitize-html version
//             that allows certain dangerous attributes/tags through
const sanitizeHtml = require('sanitize-html');

// VULNERABLE: old sanitize-html@1.4.2 does not strip all XSS vectors,
// e.g. <img src=x onerror="alert(1)"> passes through in some configs
function renderProfile(username, bio) {
  const sanitized = sanitizeHtml(bio || '');
  return `
    <html>
      <body>
        <h1>Profile: ${username}</h1>
        <div class="bio">${sanitized}</div>
      </body>
    </html>
  `;
}

// VULNERABLE: username itself is also not escaped
function renderComment(author, comment) {
  return `<div><strong>${author}</strong>: ${comment}</div>`;
}

module.exports = { renderProfile, renderComment };
